cli/src/main/java/com/devonfw/tools/ide/tool/LocalToolCommandlet.java

```diff
@@ -0,0 +1,246 @@
+
+    return installInRepo(version, edition, this.context.getDefaultToolRepository());
+  }
+
+  /**
+   * Performs the installation of the {@link #getName() tool} managed by this {@link Commandlet} only in the central
+   * software repository without touching the IDE installation.
+   *
+   * @param version the {@link VersionIdentifier} requested to be installed. May also be a
+   *        {@link VersionIdentifier#isPattern() version pattern}.
+   * @param edition the specific edition to install.
+   * @param toolRepository the {@link ToolRepository} to use.
+   * @return the {@link ToolInstallation} in the central software repository matching the given {@code version}.
+   */
+  public ToolInstallation installInRepo(VersionIdentifier version, String edition, ToolRepository toolRepository) {
+
+    VersionIdentifier resolvedVersion = toolRepository.resolveVersion(this.tool, edition, version);
+    Path toolPath = this.context.getSoftwareRepositoryPath().resolve(toolRepository.getId()).resolve(this.tool)
+        .resolve(edition).resolve(resolvedVersion.toString());
+    Path toolVersionFile = toolPath.resolve(IdeContext.FILE_SOFTWARE_VERSION);
+    FileAccess fileAccess = this.context.getFileAccess();
+    if (Files.isDirectory(toolPath)) {
+      if (Files.exists(toolVersionFile)) {
+        this.context.debug("Version {} of tool {} is already installed at {}", resolvedVersion,
+            getToolWithEdition(this.tool, edition), toolPath);
+        return createToolInstallation(toolPath, resolvedVersion, toolVersionFile);
+      }
+      this.context.warning("Deleting corrupted installation at {}", toolPath);
+      fileAccess.delete(toolPath);
+    }
+    Path target = toolRepository.download(this.tool, edition, resolvedVersion);
+    fileAccess.mkdirs(toolPath.getParent());
+    extract(target, toolPath);
+    try {
+      Files.writeString(toolVersionFile, resolvedVersion.toString(), StandardOpenOption.CREATE_NEW);
+    } catch (IOException e) {
+      throw new IllegalStateException("Failed to write version file " + toolVersionFile, e);
+    }
+    return createToolInstallation(toolPath, resolvedVersion, toolVersionFile);
+  }
+
+  private ToolInstallation createToolInstallation(Path rootDir, VersionIdentifier resolvedVersion,
+      Path toolVersionFile) {
+
+    Path linkDir = getMacOsHelper().findLinkDir(rootDir);
+    Path binDir = linkDir;
+    Path binFolder = binDir.resolve(IdeContext.FOLDER_BIN);
+    if (Files.isDirectory(binFolder)) {
+      binDir = binFolder;
+    }
+    if (linkDir != rootDir) {
+      assert (!linkDir.equals(rootDir));
+      this.context.getFileAccess().copy(toolVersionFile, linkDir, FileCopyMode.COPY_FILE_OVERRIDE);
+    }
+    return new ToolInstallation(rootDir, linkDir, binDir, resolvedVersion);
+  }
+
+  /**
+   * @return the currently installed tool version or {@code null} if not found (tool not installed).
+   */
+  protected String getInstalledToolVersion() {
```

Comment:
never duplicate such code. Either move it from ToolCommandlet here or leave it in ToolCommandlet.

Reply:
I made getInstalledVersion abstract in ToolCommandlet, implemented it in LocalToolCommandlet and in GlobalToolCommandlet it just gives null back.
Another solution would be to remove it entirely from ToolCommandlet and GlobalToolCommandlet and only implement it in LocalToolCommandlet then change GetVersionCommandlet to: 
`

    ToolCommandlet commandlet = this.tool.getValue();
    if (commandlet instanceof GlobalToolCommandlet) {
      throw new UnsupportedOperationException("Not implemented yet!");
    }
    VersionIdentifier installedVersion = ((LocalToolCommandlet)commandlet).getInstalledVersion();
    if (installedVersion == null) {
      throw new CliException("Tool " + commandlet.getName() + " is not installed!", 4);
    }
    this.context.info(installedVersion.toString());
  }`

Reply:
This method should be entirely removed. I already left a comment that this was a mistake on my end from the start.
Further do not implement methods "wrong". Either we do not support `getInstalledVersion()` for `GlobalToolCommdlet`s or we do. If we do then `GlobalToolCommdlet` can not easily implement this in a generic way. It would need to run the binary and ask it for its version like this:
```
$ docker -v
Docker version 24.0.5, build ced0996
```
Then the method has to be implemented in `DockerCommandlet` accordingly.

IMHO this makes sense as e.g. `VersionGetCommandlet` should also work for global tools.

BTW: your suggestion with `instanceof` is not making use of polymorphism. The beauty of OOP is that it simplifies such things. So if the behavior of your example would actually be desired, it would still be better to simply have the following in `GlobalToolCommandlet`:
```
@Override
protected String getInstalledToolVersion() {
  throw new UnsupportedOperationException("Not implemented yet!");
}
```
and in `LocalToolCommandlet` keep the rest of the implementation.
The `instanceof` approach makes `ToolCommandlet` to know `GlobalToolCommandlet` what otherwise is not required and more elegant.

Reply:
And finally what would make sense is something like this (most probably in `ToolCommandlet`) as it allows proper reuse:
```
  protected VersionIdentifier getInstalledVersion(Path toolPath) {
    
    if (!Files.isDirectory(toolPath)) {
      this.context.debug("Tool {} not installed in {}", getName(), toolPath);
      return null;
    }
    Path toolVersionFile = toolPath.resolve(IdeContext.FILE_SOFTWARE_VERSION);
    if (!Files.exists(toolVersionFile)) {
      Path legacyToolVersionFile = toolPath.resolve(IdeContext.FILE_LEGACY_SOFTWARE_VERSION);
      if (Files.exists(legacyToolVersionFile)) {
        toolVersionFile = legacyToolVersionFile;
      } else {
        this.context.warning("Tool {} is missing version file in {}", getName(), toolVersionFile);
        return null;
      }
    }
    try {
      String version = Files.readString(toolVersionFile).trim();
      return VersionIdentifier.of(version);
    } catch (IOException e) {
      throw new IllegalStateException("Failed to read file " + toolVersionFile, e);
    }
  }
```
Now we can easily implement this making reuse:
```
  public VersionIdentifier getInstalledVersion() {

    return getInstalledVersion(getToolPath());
  }
```

The benefit is that we can later also reuse `getInstalledVersion(Path)` to get the version of a tool from the software repository (e.g. to verify that all is sane).

Reply:
Done. Thanks for the clarification.